File: sdk/js/src/modules/realtime.ts
// sdk/js/src/modules/realtime.ts

import { io, Socket } from "socket.io-client";
import { BaseFetch } from "../utils/fetch";
import type { RealtimeOptions, RealtimeEventType } from "../types/filters";
import type { RealtimeEvent } from "../types/index";

// ─── Internal types ───────────────────────────────────────────────────────────

type EventListener<T extends Record<string, unknown> = Record<string, unknown>> = (
  event: RealtimeEvent<T>
) => void;

interface SubscribeResponse {
  channel: string;
  resource: string;
  event_types: RealtimeEventType[];
  connection_url: string;
}

interface Subscription {
  channel: string;
  resource: string;
  listeners: Set<EventListener>;
}

// ─── RealtimeModule ───────────────────────────────────────────────────────────

export class RealtimeModule {
  private http: BaseFetch;
  private projectId: string;
  private baseUrl: string;

  /** Lazily initialised Socket.io connection */
  private socket: Socket | null = null;

  /** channel name → subscription state */
  private subscriptions = new Map<string, Subscription>();

  /** Whether connect() has been called */
  private connecting = false;

  constructor(http: BaseFetch, projectId: string, baseUrl: string) {
    this.http = http;
    this.projectId = projectId;
    this.baseUrl = baseUrl;
  }

  // ─── Public API ──────────────────────────────────────────────────────────────

  /**
   * Subscribe to changes on a SQL table (INSERT, UPDATE, DELETE).
   * Returns an unsubscribe function — call it to stop listening.
   *
   * @example
   * const unsub = baas.realtime.on("posts", (event) => {
   *   console.log(event.type, event.record)
   * })
   * unsub() // stop listening
   *
   * @example — filter event types
   * const unsub = baas.realtime.on("posts", handler, { eventTypes: ["INSERT"] })
   */
  on<T extends Record<string, unknown> = Record<string, unknown>>(
    table: string,
    listener: EventListener<T>,
    options?: RealtimeOptions
  ): () => void {
    return this._subscribe(table, listener as EventListener, options);
  }

  /**
   * Subscribe to changes on a NoSQL collection (INSERT, UPDATE, DELETE).
   * Functionally identical to `.on()` — provided as a semantic alias.
   *
   * @example
   * const unsub = baas.realtime.onCollection("articles", (event) => {
   *   console.log(event.type, event.record)
   * })
   */
  onCollection<T extends Record<string, unknown> = Record<string, unknown>>(
    collection: string,
    listener: EventListener<T>,
    options?: RealtimeOptions
  ): () => void {
    return this._subscribe(collection, listener as EventListener, options);
  }

  /**
   * Disconnect from the realtime server and remove all listeners.
   * Call this when your app unmounts or the user signs out.
   */
  disconnect(): void {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
    this.subscriptions.clear();
    this.connecting = false;
  }

  /**
   * Returns true if the Socket.io connection is currently active.
   */
  isConnected(): boolean {
    return this.socket?.connected ?? false;
  }

  // ─── Internal ────────────────────────────────────────────────────────────────

  /**
   * Core subscribe logic:
   * 1. Ask the backend for the channel name via POST /subscribe
   * 2. Ensure the Socket.io connection is open
   * 3. Join the channel and route events to listeners
   */
  private _subscribe(
    resource: string,
    listener: EventListener,
    options?: RealtimeOptions
  ): () => void {
    // Fire-and-forget async setup; return unsubscribe immediately
    this._setupSubscription(resource, listener, options).catch((err) => {
      console.error(`[YourBaaS Realtime] Failed to subscribe to "${resource}":`, err);
    });

    // Return unsubscribe — works even before the async setup completes
    return () => this._removeListener(resource, listener);
  }

  private async _setupSubscription(
    resource: string,
    listener: EventListener,
    options?: RealtimeOptions
  ): Promise<void> {
    // 1. Get channel name from backend
    const res = await this.http.post<SubscribeResponse>(
      `/v1/realtime/${this.projectId}/subscribe`,
      {
        table_or_collection: resource,
        event_types: options?.eventTypes ?? ["INSERT", "UPDATE", "DELETE"],
      }
    );

    const { channel } = res.data;

    // 2. Track the subscription
    if (!this.subscriptions.has(channel)) {
      this.subscriptions.set(channel, {
        channel,
        resource,
        listeners: new Set(),
      });
    }
    this.subscriptions.get(channel)!.listeners.add(listener);

    // 3. Connect if not already connected
    await this._ensureConnected();

    // 4. Join the Socket.io room / channel
    this.socket!.emit("subscribe", { channel });

    // 5. Register the channel handler once (idempotent guard via Map)
    this._registerChannelHandler(channel);
  }

  private async _ensureConnected(): Promise<void> {
    if (this.socket?.connected) return;

    if (this.connecting) {
      // Wait for the pending connection
      await this._waitForConnection();
      return;
    }

    this.connecting = true;

    return new Promise((resolve, reject) => {
      this.socket = io(this.baseUrl, {
        path: "/socket.io",
        transports: ["websocket", "polling"],
        namespace: `/${this.projectId}`,
        auth: (cb) => {
          // Inject the API key as the connection credential
          cb({ apiKey: this._getApiKey() });
        },
        reconnection: true,
        reconnectionAttempts: 5,
        reconnectionDelay: 1000,
      });

      this.socket.once("connect", () => {
        this.connecting = false;
        resolve();
      });

      this.socket.once("connect_error", (err) => {
        this.connecting = false;
        reject(new Error(`Realtime connection failed: ${err.message}`));
      });
    });
  }

  private _waitForConnection(): Promise<void> {
    return new Promise((resolve) => {
      const interval = setInterval(() => {
        if (!this.connecting) {
          clearInterval(interval);
          resolve();
        }
      }, 50);
    });
  }

  /**
   * Register a socket event handler for a channel (only once per channel).
   * Uses the socket's own listener count to guard against duplicates.
   */
  private _registerChannelHandler(channel: string): void {
    if (!this.socket) return;

    // Avoid double-registering — socket.io counts listeners per event name
    if (this.socket.listeners(channel).length > 0) return;

    this.socket.on(channel, (raw: unknown) => {
      const sub = this.subscriptions.get(channel);
      if (!sub || sub.listeners.size === 0) return;

      const event = this._parseEvent(raw, sub.resource);
      if (!event) return;

      for (const listener of sub.listeners) {
        try {
          listener(event);
        } catch {
          // Never let a listener crash the realtime loop
        }
      }
    });
  }

  private _removeListener(resource: string, listener: EventListener): void {
    // Find the channel for this resource
    for (const [channel, sub] of this.subscriptions) {
      if (sub.resource === resource) {
        sub.listeners.delete(listener);

        // If no listeners remain, unsubscribe from the channel
        if (sub.listeners.size === 0) {
          this.subscriptions.delete(channel);
          this.socket?.emit("unsubscribe", { channel });
          this.socket?.off(channel);
        }
        return;
      }
    }
  }

  private _parseEvent(
    raw: unknown,
    resource: string
  ): RealtimeEvent | null {
    if (!raw || typeof raw !== "object") return null;

    const r = raw as Record<string, unknown>;
    const type = (r["type"] as string)?.toUpperCase() as RealtimeEventType;

    if (!["INSERT", "UPDATE", "DELETE"].includes(type)) return null;

    return {
      type,
      record: (r["new"] ?? r["record"]) as Record<string, unknown> | undefined,
      old: r["old"] as Record<string, unknown> | undefined,
      resource,
      projectId: this.projectId,
    };
  }

  /**
   * Extract the raw API key from the Authorization header injected by BaseFetch.
   * Used to authenticate the Socket.io connection.
   */
  private _getApiKey(): string {
    // BaseFetch stores the key internally; we reach it via a symbol trick.
    // In practice the BaasClient passes apiKey directly to RealtimeModule.
    return (this as unknown as { _apiKey?: string })._apiKey ?? "";
  }

  /**
   * Called by BaasClient to inject the API key for Socket.io auth.
   * @internal
   */
  _setApiKey(apiKey: string): void {
    (this as unknown as { _apiKey: string })._apiKey = apiKey;
  }
}

File: sdk/python/src/baas/modules/realtime.py
# sdk/python/src/yourbaas/modules/realtime.py
from __future__ import annotations

import logging
from typing import Any, Callable

from baas.types import RealtimeEvent, UnsubscribeFn
from baas.utils.errors import BaasError
from baas.utils.http import HttpClient

logger = logging.getLogger(__name__)


class RealtimeModule:
    """
    Subscribe to real-time database changes via Socket.io.

    Uses python-socketio's async client. The Socket.io server runs on the
    same base URL under the namespace /{project_id}.

    Usage:
        # SQL table changes
        unsub = baas.realtime.on("posts", lambda event: print(event))

        # NoSQL collection changes
        unsub2 = baas.realtime.on_collection("articles", lambda event: print(event))

        # Clean up
        unsub()
        await baas.realtime.disconnect()

    Note: Requires the `python-socketio[asyncio-client]` extra.
    """

    def __init__(self, http: HttpClient, project_id: str, base_url: str, api_key: str) -> None:
        self._http = http
        self._project_id = project_id
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._sio: Any = None
        self._connected = False
        self._sql_handlers: dict[str, list[Callable[[RealtimeEvent], None]]] = {}
        self._nosql_handlers: dict[str, list[Callable[[RealtimeEvent], None]]] = {}

    def _get_sio(self) -> Any:
        if self._sio is None:
            try:
                import socketio  # type: ignore[import]
            except ImportError as exc:
                raise BaasError(
                    code="MISSING_DEPENDENCY",
                    message=(
                        "python-socketio is required for realtime support. "
                        "Install it with: pip install 'yourbaas[realtime]'"
                    ),
                ) from exc
            self._sio = socketio.AsyncClient(logger=False, engineio_logger=False)
            self._register_events()
        return self._sio

    def _register_events(self) -> None:
        sio = self._sio

        @sio.event
        async def connect() -> None:  # type: ignore[misc]
            self._connected = True
            logger.debug("Realtime connected for project %s", self._project_id)

        @sio.event
        async def disconnect() -> None:  # type: ignore[misc]
            self._connected = False
            logger.debug("Realtime disconnected")

        @sio.on("db_change")  # type: ignore[misc]
        async def on_db_change(data: dict[str, Any]) -> None:
            resource = data.get("resource", "")
            event = RealtimeEvent(
                type=data.get("type", "INSERT"),
                record=data.get("new") or data.get("payload"),
                old_record=data.get("old"),
                table=resource,
            )
            for handler in self._sql_handlers.get(resource, []):
                try:
                    handler(event)
                except Exception:
                    logger.exception("Error in SQL realtime handler for %s", resource)

        @sio.on("collection_change")  # type: ignore[misc]
        async def on_collection_change(data: dict[str, Any]) -> None:
            resource = data.get("resource", "")
            event = RealtimeEvent(
                type=data.get("type", "INSERT"),
                record=data.get("new") or data.get("payload"),
                old_record=data.get("old"),
                collection=resource,
            )
            for handler in self._nosql_handlers.get(resource, []):
                try:
                    handler(event)
                except Exception:
                    logger.exception("Error in NoSQL realtime handler for %s", resource)

    async def _ensure_connected(self) -> None:
        sio = self._get_sio()
        if not self._connected:
            user_token = self._http.get_user_token()
            auth: dict[str, str] = {"api_key": self._api_key}
            if user_token:
                auth["token"] = user_token
            await sio.connect(
                self._base_url,
                namespaces=[f"/{self._project_id}"],
                auth=auth,
                transports=["websocket"],
            )

    def on(
        self,
        table: str,
        callback: Callable[[RealtimeEvent], None],
    ) -> UnsubscribeFn:
        """
        Subscribe to SQL table changes (INSERT, UPDATE, DELETE).

        Returns an unsubscribe callable that removes this specific listener.
        Note: Does not disconnect the socket — call disconnect() explicitly.
        """
        self._sql_handlers.setdefault(table, []).append(callback)

        def unsub() -> None:
            handlers = self._sql_handlers.get(table, [])
            try:
                handlers.remove(callback)
            except ValueError:
                pass

        return unsub

    def on_collection(
        self,
        collection: str,
        callback: Callable[[RealtimeEvent], None],
    ) -> UnsubscribeFn:
        """
        Subscribe to MongoDB collection changes (INSERT, UPDATE, DELETE).

        Returns an unsubscribe callable.
        """
        self._nosql_handlers.setdefault(collection, []).append(callback)

        def unsub() -> None:
            handlers = self._nosql_handlers.get(collection, [])
            try:
                handlers.remove(callback)
            except ValueError:
                pass

        return unsub

    async def connect(self) -> None:
        """Explicitly connect the Socket.io client."""
        await self._ensure_connected()

    async def disconnect(self) -> None:
        """Disconnect the Socket.io client and clear all handlers."""
        self._sql_handlers.clear()
        self._nosql_handlers.clear()
        if self._sio and self._connected:
            await self._sio.disconnect()
        self._connected = False
